File: sdks/csharp/examples/RateLimitingExample.cs
using System;
using System.Threading.Tasks;
using TollMesh.Cache;

class Program
{
    static async Task Main(string[] args)
    {
        var config = new ClientConfig { Host = "localhost", Port = 8080 };

        using (var client = new Client(config))
        {
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("TollMeshCache - Rate Limiting Example (C#)");
            Console.WriteLine(new string('=', 60));

            try
            {
                // Example 1: Basic rate limiting
                Console.WriteLine("\n1. Basic Rate Limiting (100 req/min)");
                Console.WriteLine(new string('-', 60));

                for (int i = 0; i < 3; i++)
                {
                    var result = await client.ConsumeAsync(
                        "user-123", 100, TimeSpan.FromMinutes(1));

                    Console.WriteLine($"Request {i + 1}:");
                    Console.WriteLine($"  Status: {(result.Ok ? "ALLOWED" : "LIMITED")}");
                    Console.WriteLine($"  Remaining: {result.Remaining}");
                }

                // Example 2: Tier-based
                Console.WriteLine("\n2. Tier-Based Rate Limiting");
                Console.WriteLine(new string('-', 60));

                var tiers = new[] { ("free", 10), ("pro", 100), ("enterprise", 1000) };
                foreach (var (tier, limit) in tiers)
                {
                    var result = await client.ConsumeAsync(
                        $"user-tier-{tier}", limit, TimeSpan.FromMinutes(1));

                    var status = result.Ok ? "✓ OK" : "✗ LIMITED";
                    Console.WriteLine($"{tier.ToUpper().PadRight(12)}: {status} ({result.Remaining} remaining)");
                }

                // Example 3: Health check
                Console.WriteLine("\n3. Server Health");
                Console.WriteLine(new string('-', 60));

                var health = await client.HealthAsync();
                Console.WriteLine($"Status: {health.Status}");
                Console.WriteLine($"Node: {health.Node}");
                Console.WriteLine($"Peers: {health.Peers}");
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
            }
            finally
            {
                Console.WriteLine("\n" + new string('=', 60));
                Console.WriteLine("Example complete!");
            }
        }
    }
}

File: sdks/java/examples/RateLimitingExample.java
package com.tollmesh.store.examples;

import com.tollmesh.store.*;
import java.time.Duration;

public class RateLimitingExample {
    public static void main(String[] args) {
        ClientConfig config = new ClientConfig()
            .setHost("localhost")
            .setPort(8080);

        try (Client client = new Client(config)) {
            System.out.println("=".repeat(60));
            System.out.println("TollMeshCache - Rate Limiting Example");
            System.out.println("=".repeat(60));

            // Example 1: Basic rate limiting
            System.out.println("\n1. Basic Rate Limiting (100 requests per minute)");
            System.out.println("-".repeat(60));

            for (int i = 0; i < 3; i++) {
                ConsumeResult result = client.consume(
                    "user-123",
                    100,
                    Duration.ofMinutes(1)
                );

                System.out.println("Request " + (i + 1) + ":");
                System.out.println("  Status: " + (result.isOk() ? "ALLOWED" : "LIMITED"));
                System.out.println("  Remaining: " + result.getRemaining());
            }

            // Example 2: API tier rate limiting
            System.out.println("\n2. Tier-Based Rate Limiting");
            System.out.println("-".repeat(60));

            String[] tiers = {"free", "pro", "enterprise"};
            int[] limits = {10, 100, 1000};

            for (int i = 0; i < tiers.length; i++) {
                ConsumeResult result = client.consume(
                    "user-tier-" + tiers[i],
                    limits[i],
                    Duration.ofMinutes(1)
                );

                String status = result.isOk() ? "✓ OK" : "✗ LIMITED";
                System.out.printf("%s: %s (%d remaining)%n",
                    String.format("%-12s", tiers[i].toUpperCase()),
                    status,
                    result.getRemaining()
                );
            }

            // Example 3: Health check
            System.out.println("\n3. Server Health Status");
            System.out.println("-".repeat(60));

            HealthResponse health = client.health();
            System.out.println("Status: " + health.getStatus());
            System.out.println("Node: " + health.getNode());
            System.out.println("Peers: " + health.getPeers());

            System.out.println("\n" + "=".repeat(60));
            System.out.println("Example complete!");
            System.out.println("=".repeat(60));

        } catch (Exception e) {
            System.err.println("ERROR: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
